velox/functions/sparksql/SplitFunctions.cpp

```diff
@@ -16,124 +16,244 @@
-      std::vector<VectorPtr>& args,
-      const TypePtr& /* outputType */,
-      exec::EvalCtx& context,
-      VectorPtr& result) const override {
-    auto delimiterVector = args[1]->as<ConstantVector<StringView>>();
-    VELOX_CHECK(
-        delimiterVector, "Split function supports only constant delimiter");
-    auto patternString = args[1]->as<ConstantVector<StringView>>()->valueAt(0);
-    VELOX_CHECK_EQ(
-        patternString.size(),
-        1,
-        "split only supports only single-character pattern");
-    char pattern = patternString.data()[0];
-    SplitCharacter splitCharacter(pattern);
-    splitCharacter.apply(rows, args, nullptr, context, result);
+  // Split with a non-empty delimiter. If limit > 0, The resulting array's
+  // length will not be more than limit and the resulting array's last entry
+  // will contain all input beyond the last matched regex. If limit <= 0,
+  // delimiter will be applied as many times as possible, and the resulting
+  // array can be of any size.
+  void split(
+      const StringView input,
+      const StringView delim,
+      int32_t limit,
+      vector_size_t row,
+      exec::VectorWriter<Array<Varchar>>& resultWriter) const {
+    VELOX_USER_CHECK(!delim.empty(), "Non-empty delimiter is expected");
+    // Add new array (for the new row) to our array vector.
+    resultWriter.setOffset(row);
+    auto& arrayWriter = resultWriter.current();
+
+    // Trivial case of converting string to array with 1 element.
+    if (limit == 1) {
+      arrayWriter.add_item().setNoCopy(input);
+      resultWriter.commit();
+      return;
+    }
+
+    // We walk through our input cutting off the pieces using the delimiter and
+    // add them to the elements vector, until we reach the end of the
+    // string or the limit.
+    int32_t addedElements{0};
+    auto* re = cache_.findOrCompile(delim);
+    const size_t end = input.size();
+    const char* start = input.data();
+    const auto re2String = re2::StringPiece(start, end);
+    size_t pos = 0;
+
+    re2::StringPiece subMatches[1];
+    // Matches a regular expression against a portion of the input string,
+    // starting from 'pos' to the end of the input string. The match is not
+    // anchored, which means it can start at any position in the string. If a
+    // match is found, the matched portion of the string is stored in
+    // 'subMatches'. The '1' indicates that we are only interested in the first
+    // match found from the current position 'pos' in each iteration of the
+    // loop.
+    while (re->Match(
+        re2String, pos, end, RE2::Anchor::UNANCHORED, subMatches, 1)) {
+      const auto fullMatch = subMatches[0];
+      auto offset = fullMatch.data() - start;
+      const auto size = fullMatch.size();
+      if (offset >= end) {
+        break;
+      }
+
+      // Hit an empty match for the specific case when the delimiter has '|' as
+      // a suffix. For example, if the string is 'abc' and the delimiter is 'd|'
```

Comment:
Is it a general handling for empty match or just for the case when `delimiter has '|' as suffix` as you mentioned? By reading the code, I assume it is a general handling, then we could make the comment more concise. Below is some example:

> When hitting an empty match, split the character at the current 'pos' and put it into the result array, followed by an empty tail string at last, e.g., the result array for split('abc', 'd|') is ["a", "b", "c", ""].